src/Microsoft.AspNetCore.Mvc.Core/Internal/ObjectMethodExecutor.cs

```diff
@@ -87,5 +103,151 @@ private static ActionExecutor WrapVoidAction(VoidActionExecutor executor)
+                var voidExecutor = lambda.Compile();
+                return WrapVoidActionAsync(voidExecutor);
+            }
+            else
+            {
+                // must coerce methodCall to match ActionExecutorAsync signature
+                var coerceMethodCall = GetCoerceMethodCallExpression(methodCall, methodInfo);
+                var lambda = Expression.Lambda<ActionExecutorAsync>(coerceMethodCall, targetParameter, parametersParameter);
+                return lambda.Compile();
+            }
+        }
+
+        // We need to CoerceResult as the object value returned from methodInfo.Invoke has to be cast to a Task<T>.
+        // This is necessary to enable calling await on the returned task.
+        // i.e we need to write the following var result = await (Task<ActualType>)mInfo.Invoke.
+        // Returning Task<object> enables us to await on the result.
+        private static Expression GetCoerceMethodCallExpression(MethodCallExpression methodCall, MethodInfo methodInfo)
+        {
+            var castMethodCall = Expression.Convert(methodCall, typeof(object));
+            Type returnType = methodCall.Type;
+
+            if (typeof(Task).IsAssignableFrom(returnType))
+            {
+                if (returnType == typeof(Task))
+                {
+                    Expression<Func<object, string, Type, Task<object>>> coerceTaskExpression = ((result, methodName, declaringType) => ObjectMethodExecutor.CoerceTaskType(result, methodName, declaringType));
+                    var coerceMethodInfo = ((MethodCallExpression)coerceTaskExpression.Body).Method;
+                    var stringExpression = Expression.Constant(methodInfo.Name);
+                    var typeExpression = Expression.Constant(methodInfo.DeclaringType);
+                    return Expression.Call(null, coerceMethodInfo, castMethodCall, stringExpression, typeExpression);
+                }
+
+                var taskValueType = GetTaskInnerTypeOrNull(returnType);
+                if (taskValueType != null)
+                {
+                    // for: public Task<T> Action()
+                    // constructs: return (Task<object>)Convert<T>((Task<T>)result)
+                    var genericMethodInfo = _convertOfTMethod.MakeGenericMethod(taskValueType);
+                    var genericMethodCall = Expression.Call(null, genericMethodInfo, castMethodCall);
+                    var convertedResult = Expression.Convert(genericMethodCall, typeof(Task<object>));
+                    return convertedResult;
+                }
+
+                // This will be the case for types which have derived from Task and Task<T>
+                throw new InvalidOperationException(Resources.FormatActionExecutor_UnexpectedTaskInstance(
+                    methodInfo.Name,
+                    methodInfo.DeclaringType));
+            }
+
+            // Action methods which use dynamic keyword but return a Task or Task<T> will also be coerced to Task<object>
```

Comment:
I'd add more to this comment (or delete it). This gives the impression that `dynamic` is somehow special, it's not. It behaves just like `object`.